jbmc/src/java_bytecode/java_bytecode_parser.cpp

```diff
@@ -1640,6 +1640,48 @@ void java_bytecode_parsert::rclass_attribute(classt &parsed_class)
     parsed_class.attribute_bootstrapmethods_read = true;
     read_bootstrapmethods_entry(parsed_class);
   }
+  else if(attribute_name == "InnerClasses")
+  {
+    u2 number_of_classes = read_u2();
+    u4 number_of_bytes_to_be_read = number_of_classes * 8 + 2;
+    INVARIANT(
+      number_of_bytes_to_be_read == attribute_length,
+      "The number of bytes to be read for the InnerClasses attribute does not "
+      "match the attribute length.");
+
+    const std::function<pool_entryt &(u2)> pool_entry_lambda =
+      [this](u2 index) -> pool_entryt & { return pool_entry(index); };
+    std::function<std::string(std::string, char)> remove_separator_char =
+      [](std::string str, char ch) {
+        str.erase(std::remove(str.begin(), str.end(), ch), str.end());
+        return str;
+      };
+
+    for(int i = 0; i < number_of_classes; i++)
+    {
+      u2 inner_class_info_index = read_u2();
+      UNUSED u2 outer_class_info_index = read_u2();
+      UNUSED u2 inner_name_index = read_u2();
+      u2 inner_class_access_flags = read_u2();
+
+      if(inner_class_info_index != 0)
+      {
+        std::string inner_class_info_name =
+          class_infot(pool_entry(inner_class_info_index))
+            .get_name(pool_entry_lambda);
+        UNUSED bool is_private = inner_class_access_flags & ACC_PRIVATE;
+        UNUSED bool is_public = inner_class_access_flags & ACC_PUBLIC;
+        UNUSED bool is_protected = inner_class_access_flags & ACC_PROTECTED;
+
+        // If the original parsed class name matches the inner class name
+        // the parsed class is an inner class, so overwrite the parsed class'
```

Comment:
Why does this work? Also possibly I think you might be able to check whether `outer_class_info_index == 0`? 

> If C is not a member of a class or an interface (that is, if C is a top-level class or interface (JLS §7.6) or a local class (JLS §14.3) or an anonymous class (JLS §15.9.5)), the value of the outer_class_info_index item must be zero.

Though that sounds like anonymous and local classes would be marked as not inner, which I guess is true?

Reply:
Anonymous classes and local classes are both inner classes. If they are not being marked as such then we have a problem. I think the accessibility for these should always be private.

Reply:
Thinking about this, it may be advantageous to have a is_anonymous / is_local flag set at this point. We currently have an `is_anonymous` method in `state_typet`, which attempts to work this out based on regexing the name, which isn't the ideal approach.

Reply:
The problems on test-gen were related to treating anonymous/local classes the same as inner classes. This has been updated by checking that outer_class_info_index  is 0 or not (it's 0 when it is outer, interface, local or anonymous). @thk123 The reason we can't just check outer_class_info_index is because the we only want to overwrite the accessibility information when the current parsed class (not the current parsed inner class information) is an inner class. As far as I can tell, there is no magic way to do this from the spec.